tests/turnstile_chain.c

```diff
@@ -0,0 +1,106 @@
+#include <ktest.h>
+#include <klog.h>
+#include <mutex.h>
+#include <runq.h>
+#include <sched.h>
+#include <thread.h>
+
+// length of the propagation chain
+#define T 4
+
+// TODO replace asserts about priority borrowing with some accessor
+/* TODO consider where we could check restoring priorities
+ * I think that all threads should keep HIGH priority until the last propagator
+ * has acquired 2nd mutex and then the priorities would be restored to initial
+ * ones (propagator_priority(i) for propagator[i]).
+ */
+
+static mtx_t mtx[T + 1];
+// TODO consider whether this would be any useful
+// volatile int mtx_passed[T + 1];
+
+// for simpler code: propagator[0] = starter
+static thread_t *propagator[T + 1];
+static thread_t *starter;
+
+static void set_base_prio(thread_t *td, prio_t prio) {
+  WITH_SPINLOCK(td->td_spin) {
+    sched_set_prio(td, prio);
+  }
+}
+
+// n <- [0..T]
+// TODO maybe rename to separate_prio or something?
+static int propagator_prio(int n) {
+  // TODO should this remain `n + 1` and not just `n`?
+  return (n + 1) * RQ_PPQ;
+}
+
+/* n <- [1..T]
+ * nth propagator will contest nth mutex and block on (n-1)th mutex (owned
+ * by (n-1)th propagator) causing priority propagation to [0..n-1]th propagator
+ */
+static void propagator_routine(void *argn) {
+  int n = (int)argn;
+
+  assert(thread_self()->td_prio == propagator_prio(n));
+  assert(MTX_OWNER(&mtx[n]) == NULL);
+  WITH_MTX_LOCK (&mtx[n]) {
+
+    assert(MTX_OWNER(&mtx[n - 1]) == propagator[n - 1]);
+    WITH_MTX_LOCK (&mtx[n - 1]) {
+      // nothing here
+    }
+  }
+  assert(thread_self()->td_prio == propagator_prio(n));
+}
+
+static void starter_routine(void *_arg) {
+  assert(MTX_OWNER(&mtx[0]) == NULL);
+  WITH_MTX_LOCK (&mtx[0]) {
+
+    for (int i = 1; i <= T; i++) {
+      WITH_NO_PREEMPTION {
+        set_base_prio(propagator[i], propagator_prio(i));
+        sched_add(propagator[i]);
+        assert(thread_self()->td_prio == propagator_prio(i - 1));
+      }
+
+      // check if the priorities have propagated correctly
+      for (int j = 0; j < i; j++) {
```

Comment:
I'd add assert between for loops: `assert(TD_IS_LOCKED(propagator[i]));`.

Reply:
Or add a comment that `propagator[i]` is locked on `mtx[n-1]` and when turnstile PR comes in, we'll change it to appropriate assert.